### PR TITLE
Further refine the registration and verification views.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/view controllers/CodeVerificationViewController.m [hunk 1]
```diff
@@ -392,18 +392,6 @@ - (void)dismissKeyboardFromAppropriateSubView {
     [self.view endEditing:NO];
 }
 
-- (NSString *)stringByFilteringString:(NSString *)input
```
Comment (charlesmchen) [inline Signal/src/view controllers/CodeVerificationViewController.m]: StringUtils turns out to have a `digitsOnly` method which is simpler.

File: Signal/src/view controllers/CodeVerificationViewController.m [hunk 2]
```diff
@@ -474,9 +458,17 @@ - (BOOL)textFieldShouldReturn:(UITextField *)textField {
...
+    NSString *rawNewText = verificationCode.digitsOnly;
+    NSString *formattedNewText = (rawNewText.length <= 3
+                                  ? rawNewText
+                                  : [[[rawNewText substringToIndex:3]
+                                      stringByAppendingString:@"-"]
+                                     stringByAppendingString:[rawNewText substringFromIndex:3]]);
+    self.challengeTextField.text = formattedNewText;
+    // Move the cursor after the newly inserted text.
+    UITextPosition *newPosition = [self.challengeTextField endOfDocument];
+    self.challengeTextField.selectedTextRange = [self.challengeTextField textRangeFromPosition:newPosition
```
Comment (charlesmchen) [inline Signal/src/view controllers/CodeVerificationViewController.m]: If users open this view to verify a code received by side channel, we weren't formatting it properly.

File: Signal/src/view controllers/RegistrationViewController.m [hunk 3]
```diff
@@ -148,48 +148,60 @@ - (void)dismissKeyboardFromAppropriateSubView {
...
+    // * Invalid input should be simply ignored.
+    //
+    // We accomplish this by being permissive and trying to "take as much of the user
+    // input as possible".
+    //
+    // * Always accept deletes.
+    // * Ignore invalid input.
+    // * Take partial input if possible.
+    
+    NSString *oldText = textField.text;
```
Comment (charlesmchen) [inline Signal/src/view controllers/RegistrationViewController.m]: I aligned this logic with the filtering and auto-formatting logic in the code verification view.

File: Signal/src/view controllers/RegistrationViewController.m [hunk 4]
```diff
@@ -148,48 +148,60 @@ - (void)dismissKeyboardFromAppropriateSubView {
...
+    [textField positionFromPosition:textField.beginningOfDocument offset:(NSInteger)cursorPositionAfterReformat];
     [textField setSelectedTextRange:[textField textRangeFromPosition:pos toPosition:pos]];
-
+    
     return NO; // inform our caller that we took care of performing the change
 }
 
+- (BOOL)textFieldShouldReturn:(UITextField *)textField {
+    [self sendCodeAction:nil];
+    [textField resignFirstResponder];
```
Comment (charlesmchen) [inline Signal/src/view controllers/RegistrationViewController.m]: The return key now submits the phone number which will make it easier for developers to register in a simulator.

File: Signal/src/view controllers/CodeVerificationViewController.m [hunk 5]
```diff
@@ -139,6 +139,7 @@ - (void)createViews {
                                                         @"Text field placeholder for SMS verification code during registration");
     _challengeTextField.font = [UIFont ows_lightFontWithSize:21.f];
     _challengeTextField.textAlignment = NSTextAlignmentCenter;
+    _challengeTextField.keyboardType = UIKeyboardTypePhonePad;
```
Comment (charlesmchen) [inline Signal/src/view controllers/CodeVerificationViewController.m]: The "registration" view used the "decimal digits" keyboard whereas the "code verification" view used the conventional keyboard.  Both values should only contain decimal digits.  Now both use the "decimal digits" keypad.

I've verified that the layout still works on our largest and smallest form factors.

screen shot 2017-02-16 at 9 37 31 pm
screen shot 2017-02-16 at 9 37 16 pm

File: Signal/src/view controllers/RegistrationViewController.m (hunk 3 above)
Comment (michaelkirk) [inline Signal/src/view controllers/RegistrationViewController.m]: nit: Are the changes to this method fixing any known problem? Even though your implementation is arguably cleaner, it seems like code that would be easy to get wrong. If it ain't broke, yadda yadda.

File: Signal/src/view controllers/RegistrationViewController.m [hunk 6]
```diff
@@ -148,48 +149,60 @@ - (void)dismissKeyboardFromAppropriateSubView {
...
+                                 stringByAppendingString:right];
+    // 5. Construct the "formatted" new text by inserting a hyphen if necessary.
     // reformat the phone number, trying to keep the cursor beside the inserted or deleted digit
-    bool isJustDeletion = string.length == 0;
+    bool isJustDeletion = insertionText.length == 0;
+    NSUInteger cursorPositionAfterChange = left.length + center.length;
     NSString *textAfterReformat =
-        [PhoneNumber bestEffortFormatPartialUserSpecifiedTextToLookLikeAPhoneNumber:textAfterChange.digitsOnly
-                                                     withSpecifiedCountryCodeString:_countryCodeButton.titleLabel.text];
+    [PhoneNumber bestEffortFormatPartialUserSpecifiedTextToLookLikeAPhoneNumber:textAfterChange
```
Comment (michaelkirk) [inline Signal/src/view controllers/RegistrationViewController.m]: nit: clang-format, old formatting was correct.

File: Signal/src/view controllers/CodeVerificationViewController.m (hunk 2 above)
Comment (michaelkirk) [inline Signal/src/view controllers/CodeVerificationViewController.m]: You mean clicking via the SMS link was broken? Or something else?

File: Signal/src/view controllers/CodeVerificationViewController.m (hunk 5 above)
Comment (michaelkirk) [inline Signal/src/view controllers/CodeVerificationViewController.m]: > the "code verification" view used the conventional keyboard. 

Yikes, that's a regression. Previously you'd see a numeric only keypad which omits the special non-numeric phone characters (+X#), which seems preferable. 

digit-pad

This was on a physical 5c running 2.7.1

File: Signal/src/view controllers/CodeVerificationViewController.m (hunk 2 above)
Comment (charlesmchen) [inline Signal/src/view controllers/CodeVerificationViewController.m]: Not broken, but the code would show up as "123456"; it should show up as "123-456".

File: Signal/src/view controllers/RegistrationViewController.m (hunk 6 above)
Comment (charlesmchen) [inline Signal/src/view controllers/RegistrationViewController.m]: Done.

File: Signal/src/view controllers/CodeVerificationViewController.m (hunk 5 above)
Comment (charlesmchen) [inline Signal/src/view controllers/CodeVerificationViewController.m]: Done.

**Pull request merged**